src/io/rsm/rsm_client.hpp

```diff
@@ -131,6 +153,117 @@ class RsmClient {
 
     return TimedOut{};
   }
+
+  /// AsyncRead methods
+  void SendAsyncReadRequest(ReadRequestT req) {
+    MG_ASSERT(!async_read_);
+
+    ReadRequest<ReadRequestT> read_req;
+    read_req.operation = req;
+
+    if (!async_read_before_) {
+      async_read_before_ = io_.Now();
+    }
+    current_read_request_ = req;
+    async_read_ = io_.template Request<ReadRequest<ReadRequestT>, ReadResponse<ReadResponseT>>(leader_, read_req);
+  }
+
+  std::optional<BasicResult<TimedOut, ReadResponseT>> PollAsyncReadRequest() {
+    MG_ASSERT(async_read_);
+
+    if (!async_read_->IsReady()) {
+      return std::nullopt;
+    }
+
+    return AwaitAsyncReadRequest();
+  }
+
+  std::optional<BasicResult<TimedOut, ReadResponseT>> AwaitAsyncReadRequest() {
+    ResponseResult<ReadResponse<ReadResponseT>> get_response_result = std::move(*async_read_).Wait();
+    async_read_.reset();
+
+    const Duration overall_timeout = io_.GetDefaultTimeout();
+    const bool past_time_out = io_.Now() < *async_read_before_ + overall_timeout;
+    const bool result_has_error = get_response_result.HasError();
+
+    if (result_has_error && past_time_out) {
+      // TODO static assert the exact type of error.
+      spdlog::debug("client timed out while trying to communicate with leader server {}", leader_.ToString());
+      async_read_before_ = std::nullopt;
+      return TimedOut{};
+    }
+    if (!result_has_error) {
+      ResponseEnvelope<ReadResponse<ReadResponseT>> &&get_response_envelope = std::move(get_response_result.GetValue());
+      ReadResponse<ReadResponseT> &&read_get_response = std::move(get_response_envelope.message);
+
+      PossiblyRedirectLeader(read_get_response);
+
+      if (read_get_response.success) {
+        async_read_before_ = std::nullopt;
+        return std::move(read_get_response.read_return);
+      }
+    } else if (result_has_error) {
+      SelectRandomLeader();
```

Comment:
If there are 3 addresses there is 33% chance to select randomly the same recipient. Can this be an issue?

Reply:
When this function is being called, it is possible that previous leader_ was in a `Candidate` state. This means that the Raft group is possibly in the process of leader election. So picking the same recipient than before might just be the correct guess.

Reply:
Then the question would be what's the probability of this happening and if we can somehow mitigate this through some metadata. But I think this is beyond the current scope -- just thinking out loud here.